src/main/java/com/pingcap/tikv/RegionManager.java

```diff
@@ -153,13 +157,51 @@ private boolean putRegion(Region region) {
             keyToRegionIdCache.put(makeRange(region.getStartKey(), region.getEndKey()),
                                              region.getId());
         } finally {
-            lock.writeLock().lock();
+            lock.writeLock().unlock();
         }
         return true;
     }
 
+    public void onRegionStale(long regionID, List<Region> regions) {
+        invalidateRegion(regionID);
+        regions.stream().map(r -> new TiRegion(r, r.getPeers(0))).forEach(this::putRegion);
+    }
+
+    private boolean isRegionLeaderSwitched(Region region, long storeID) {
```

Comment:
how this works? if a leader changed itself is actually not tested. Please rename it to 
also please change storeId to leaderPeerId.